File: jupyterqt/network/rest_client.py
import requests

from jupyterqt.config import ServerConfig


class RestClient:
    def __init__(self, config: ServerConfig):
        self._config = config
        self._session = requests.Session()
        self._session.headers.update(config.headers)
        self._fetchXsrf()

    def updateConfig(self, config: ServerConfig) -> None:
        self._config = config
        self._session = requests.Session()
        self._session.headers.update(config.headers)
        self._fetchXsrf()

    def _fetchXsrf(self) -> None:
        """GET /login to obtain the _xsrf cookie required for mutating requests."""
        try:
            self._session.get(f"{self._config.base_url}/login", timeout=5)
        except Exception:
            pass

    def _xsrfHeaders(self) -> dict:
        """Return X-XSRFToken header if we have the cookie, else empty dict."""
        token = self._session.cookies.get("_xsrf", "")
        return {"X-XSRFToken": token} if token else {}

    # ---------- Notebooks ----------

    def listContents(self, path: str = "") -> dict:
        url = f"{self._config.base_url}/api/contents/{path}"
        r = self._session.get(url, timeout=15)
        r.raise_for_status()
        return r.json()

    def getNotebook(self, path: str) -> dict:
        url = f"{self._config.base_url}/api/contents/{path}"
        r = self._session.get(url, params={"type": "notebook"}, timeout=30)
        r.raise_for_status()
        return r.json()

    def saveNotebook(self, path: str, content: dict) -> None:
        url = f"{self._config.base_url}/api/contents/{path}"
        payload = {"type": "notebook", "content": content}
        r = self._session.put(url, json=payload, timeout=30,
                              headers=self._xsrfHeaders())
        r.raise_for_status()

    def createDirectory(self, path: str) -> dict:
        """Create a new directory; path is the full path including the new dir name."""
        parent = path.rsplit("/", 1)[0] if "/" in path else ""
        url = f"{self._config.base_url}/api/contents/{parent}"
        r = self._session.post(url, json={"type": "directory"}, timeout=15,
                               headers=self._xsrfHeaders())
        r.raise_for_status()
        result = r.json()
        # Rename from Untitled Folder to the requested name if needed
        new_name = path.rsplit("/", 1)[-1]
        if result.get("name") != new_name:
            self.renameFile(result["path"], path)
        return result

    def renameFile(self, old_path: str, new_path: str) -> dict:
        url = f"{self._config.base_url}/api/contents/{old_path}"
        r = self._session.patch(url, json={"path": new_path}, timeout=15,
                                headers=self._xsrfHeaders())
        r.raise_for_status()
        return r.json()

    def deleteFile(self, path: str) -> None:
        url = f"{self._config.base_url}/api/contents/{path}"
        r = self._session.delete(url, timeout=15, headers=self._xsrfHeaders())
        if r.status_code not in (200, 204, 404):
            r.raise_for_status()

    def copyFile(self, path: str) -> dict:
        """Duplicate a file into the same directory."""
        parent = path.rsplit("/", 1)[0] if "/" in path else ""
        url = f"{self._config.base_url}/api/contents/{parent}"
        r = self._session.post(url, json={"copy_from": path}, timeout=15,
                               headers=self._xsrfHeaders())
        r.raise_for_status()
        return r.json()

    def createNotebook(self, path: str = "") -> dict:
        url = f"{self._config.base_url}/api/contents/{path}"
        r = self._session.post(url, json={"type": "notebook"}, timeout=15,
                               headers=self._xsrfHeaders())
        r.raise_for_status()
        return r.json()

    # ---------- Kernels ----------

    def listKernels(self) -> list[dict]:
        url = f"{self._config.base_url}/api/kernels"
        r = self._session.get(url, timeout=10)
        r.raise_for_status()
        return r.json()

    def startKernel(self, kernel_name: str = "python3") -> dict:
        url = f"{self._config.base_url}/api/kernels"
        r = self._session.post(url, json={"name": kernel_name}, timeout=30,
                               headers=self._xsrfHeaders())
        r.raise_for_status()
        return r.json()

    def shutdownKernel(self, kernel_id: str) -> None:
        url = f"{self._config.base_url}/api/kernels/{kernel_id}"
        r = self._session.delete(url, timeout=10, headers=self._xsrfHeaders())
        if r.status_code not in (200, 204, 404):
            r.raise_for_status()

    def restartKernel(self, kernel_id: str) -> dict:
        url = f"{self._config.base_url}/api/kernels/{kernel_id}/restart"
        r = self._session.post(url, timeout=30, headers=self._xsrfHeaders())
        r.raise_for_status()
        return r.json()

    def interruptKernel(self, kernel_id: str) -> None:
        url = f"{self._config.base_url}/api/kernels/{kernel_id}/interrupt"
        r = self._session.post(url, timeout=10, headers=self._xsrfHeaders())
        r.raise_for_status()

    def getKernel(self, kernel_id: str) -> dict:
        url = f"{self._config.base_url}/api/kernels/{kernel_id}"
        r = self._session.get(url, timeout=10)
        r.raise_for_status()
        return r.json()

    # ---------- Sessions ----------

    def listSessions(self) -> list[dict]:
        url = f"{self._config.base_url}/api/sessions"
        r = self._session.get(url, timeout=10)
        r.raise_for_status()
        return r.json()

    def getServerInfo(self) -> dict:
        url = f"{self._config.base_url}/api"
        r = self._session.get(url, timeout=5)
        r.raise_for_status()
        return r.json()

    def checkServer(self) -> str:
        """Returns 'ok', 'unauthorized', or an error message string."""
        try:
            url = f"{self._config.base_url}/api"
            r = self._session.get(url, timeout=5)
            if r.status_code == 200:
                return "ok"
            elif r.status_code in (401, 403):
                return "unauthorized"
            else:
                return f"HTTP {r.status_code}"
        except Exception as e:
            return f"error: {e}"

File: cpp/src/network/RestClient.cpp
#include "RestClient.h"
#include <QNetworkCookieJar>
#include <QNetworkCookie>
#include <QJsonDocument>
#include <QUrlQuery>

RestClient::RestClient(QObject* parent)
    : QObject(parent)
    , m_nam(new QNetworkAccessManager(this))
{
}

void RestClient::setConfig(const ServerConfig& config) {
    m_config = config;
}

QByteArray RestClient::xsrfToken() const {
    QUrl url(m_config.baseUrl);
    auto cookies = m_nam->cookieJar()->cookiesForUrl(url);
    for (const auto& cookie : cookies) {
        if (cookie.name() == "_xsrf") {
            return cookie.value();
        }
    }
    return {};
}

QNetworkRequest RestClient::buildRequest(const QString& endpoint) const {
    QUrl url(m_config.baseUrl + endpoint);
    QNetworkRequest req(url);
    req.setHeader(QNetworkRequest::ContentTypeHeader, QStringLiteral("application/json"));

    if (!m_config.token.isEmpty()) {
        req.setRawHeader("Authorization",
            QStringLiteral("token %1").arg(m_config.token).toUtf8());
    }

    auto xsrf = xsrfToken();
    if (!xsrf.isEmpty()) {
        req.setRawHeader("X-XSRFToken", xsrf);
    }

    return req;
}

QNetworkReply* RestClient::fetchXsrf() {
    QUrl url(m_config.baseUrl + QStringLiteral("/login"));
    QNetworkRequest req(url);
    if (!m_config.token.isEmpty()) {
        req.setRawHeader("Authorization",
            QStringLiteral("token %1").arg(m_config.token).toUtf8());
    }
    return m_nam->get(req);
}

QNetworkReply* RestClient::listContents(const QString& path) {
    QString endpoint = QStringLiteral("/api/contents");
    if (!path.isEmpty()) endpoint += QStringLiteral("/") + path;
    return m_nam->get(buildRequest(endpoint));
}

QNetworkReply* RestClient::getNotebook(const QString& path) {
    QString endpoint = QStringLiteral("/api/contents/") + path;
    return m_nam->get(buildRequest(endpoint));
}

QNetworkReply* RestClient::saveNotebook(const QString& path, const QJsonObject& content) {
    QString endpoint = QStringLiteral("/api/contents/") + path;
    auto req = buildRequest(endpoint);

    QJsonObject body;
    body[QStringLiteral("type")]    = QStringLiteral("notebook");
    body[QStringLiteral("content")] = content;

    return m_nam->put(req, QJsonDocument(body).toJson(QJsonDocument::Compact));
}

QNetworkReply* RestClient::createNotebook(const QString& dir) {
    QString endpoint = QStringLiteral("/api/contents");
    if (!dir.isEmpty()) endpoint += QStringLiteral("/") + dir;

    auto req = buildRequest(endpoint);
    QJsonObject body;
    body[QStringLiteral("type")] = QStringLiteral("notebook");

    return m_nam->post(req, QJsonDocument(body).toJson(QJsonDocument::Compact));
}

QNetworkReply* RestClient::startKernel(const QString& name) {
    auto req = buildRequest(QStringLiteral("/api/kernels"));
    QJsonObject body;
    body[QStringLiteral("name")] = name.isEmpty() ? QStringLiteral("python3") : name;
    return m_nam->post(req, QJsonDocument(body).toJson(QJsonDocument::Compact));
}

QNetworkReply* RestClient::shutdownKernel(const QString& id) {
    auto req = buildRequest(QStringLiteral("/api/kernels/") + id);
    return m_nam->deleteResource(req);
}

QNetworkReply* RestClient::restartKernel(const QString& id) {
    auto req = buildRequest(QStringLiteral("/api/kernels/") + id + QStringLiteral("/restart"));
    return m_nam->post(req, QByteArray());
}

QNetworkReply* RestClient::interruptKernel(const QString& id) {
    auto req = buildRequest(QStringLiteral("/api/kernels/") + id + QStringLiteral("/interrupt"));
    return m_nam->post(req, QByteArray());
}

QNetworkReply* RestClient::checkServer() {
    auto req = buildRequest(QStringLiteral("/api"));
    return m_nam->get(req);
}
